### PR TITLE
added request validation for urls with and without ports

### PR DESCRIPTION
**Pull request by eshanholtz**
PR for DI-322

### CODE REVIEW

**Comment by eshanholtz** [comment]
Just realized I forgot to add url builder unit tests. Adding now.

File: lib/twilio-ruby/security/request_validator.rb [hunk 1]
```diff
@@ -25,16 +25,25 @@ def initialize(auth_token = nil)
...
-        else
-          expected_signature = build_signature_for(url, {})
-          body_hash = URI.decode_www_form(URI(url).query).to_h['bodySHA256']
-          expected_hash = build_hash_for(params)
-          secure_compare(expected_signature, signature) && secure_compare(expected_hash, body_hash)
+        unless params_hash.is_a? Enumerable
+          body_hash = URI.decode_www_form(parsed_url.query).to_h['bodySHA256']
+          params_hash = build_hash_for(params)
+          valid_body = !(params_hash.nil? || body_hash.nil?) && secure_compare(params_hash, body_hash)
+          params_hash = {}
```
Comment (childish-sambino) [inline lib/twilio-ruby/security/request_validator.rb]: Noice!

File: lib/twilio-ruby/security/request_validator.rb [hunk 2]
```diff
@@ -92,6 +101,50 @@ def body_or_hash(params_or_body)
...
+      # @return [String] The URL with the standard port number
+      def build_url_with_port_for(parsed_url)
+        url = ''
+
+        url += parsed_url.scheme ? "#{parsed_url.scheme}://" : ''
+        url += parsed_url.userinfo ? "#{parsed_url.userinfo}@" : ''
+        url += parsed_url.host ? "#{parsed_url.host}:#{parsed_url.default_port}" : ''
+        url += parsed_url.path
+        url += parsed_url.query ? "?#{parsed_url.query}" : ''
+        url += parsed_url.fragment ? "##{parsed_url.fragment}" : ''
```
Comment (childish-sambino) [inline lib/twilio-ruby/security/request_validator.rb]: Don't think there's any tests for this line, but I assume it's the correct syntax.

File: lib/twilio-ruby/security/request_validator.rb [hunk 3]
```diff
@@ -92,6 +101,50 @@ def body_or_hash(params_or_body)
...
+      #
+      # @param [URI] parsed_url The parsed request url
+      #
+      # @return [String] The URL with the standard port number
+      def build_url_with_port_for(parsed_url)
+        url = ''
+
+        url += parsed_url.scheme ? "#{parsed_url.scheme}://" : ''
+        url += parsed_url.userinfo ? "#{parsed_url.userinfo}@" : ''
+        url += parsed_url.host ? "#{parsed_url.host}:#{parsed_url.default_port}" : ''
```
Comment (childish-sambino) [inline lib/twilio-ruby/security/request_validator.rb]: `default_port` looks weird here.

Comment (eshanholtz) [reply lib/twilio-ruby/security/request_validator.rb]: Well, that's not very nice. What did `default_port` ever do to you?

**Pull request merged**